pkg/login/social/socialimpl/service.go

```diff
@@ -0,0 +1,247 @@
+package socialimpl
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"net"
+	"net/http"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/grafana/grafana/pkg/infra/log"
+	"github.com/grafana/grafana/pkg/infra/remotecache"
+	"github.com/grafana/grafana/pkg/infra/usagestats"
+	"github.com/grafana/grafana/pkg/login/social"
+	"github.com/grafana/grafana/pkg/login/social/connectors"
+	"github.com/grafana/grafana/pkg/login/social/constants"
+	"github.com/grafana/grafana/pkg/login/social/models"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+	"github.com/grafana/grafana/pkg/services/ssosettings"
+	"github.com/grafana/grafana/pkg/services/supportbundles"
+	"github.com/grafana/grafana/pkg/setting"
+	"gopkg.in/ini.v1"
+)
+
+var (
+	allOauthes = []string{constants.GitHubProviderName, constants.GitlabProviderName, constants.GoogleProviderName, constants.GenericOAuthProviderName, constants.GrafanaNetProviderName,
+		constants.GrafanaComProviderName, constants.AzureADProviderName, constants.OktaProviderName}
+)
+
+type SocialService struct {
+	cfg *setting.Cfg
+
+	socialMap map[string]social.SocialConnector
+	log       log.Logger
+}
+
+func ProvideService(cfg *setting.Cfg,
+	features *featuremgmt.FeatureManager,
+	usageStats usagestats.Service,
+	bundleRegistry supportbundles.Service,
+	cache remotecache.CacheStorage,
+	ssoSettings ssosettings.Service,
+) *SocialService {
+	ss := &SocialService{
+		cfg:       cfg,
+		socialMap: make(map[string]social.SocialConnector),
+		log:       log.New("login.social"),
+	}
+
+	usageStats.RegisterMetricsFunc(ss.getUsageStats)
+
+	if features.IsEnabledGlobally(featuremgmt.FlagSsoSettingsApi) {
+		allSettings, err := ssoSettings.List(context.Background())
+		if err != nil {
+			ss.log.Error("Failed to get SSO settings", "error", err)
+		}
+
+		for _, ssoSetting := range allSettings {
+			conn, err := createOAuthConnector(ssoSetting.Provider, ssoSetting.OAuthSettings, cfg, features, cache)
+			if err != nil {
+				ss.log.Error("Failed to create OAuth provider", "error", err, "provider", ssoSetting.Provider)
+				continue
+			}
+
+			ssoSettings.RegisterReloadable(ssoSetting.Provider, conn)
+			ss.socialMap[ssoSetting.Provider] = conn
```

Comment:
I think that trying to keep the consistency between the SSO settings in the DB and the `socialMap` cache will not be easy. Wouldn't be easier to always load the SSO settings from DB whenever we need them?

I know that SSO settings may not exist in the DB for some providers, but we do have fallback that reads them from the ini file/env vars, right?

Reply:
I don't think that 3 DB queries to a tiny table during login will have a big impact on performance to justify the need of that cache. But if we get rid of the cache we will have reduced complexity. We won't need the reloadable logic and the sso settings service won't have the dependency on connectors.

I think we should try to keep things simple and define simple responsibilities for each components involved in SSO login. In my opinion:
- SSO settings service should only be responsible for saving new settings and retrieving them from the DB, ini file, env vars. It should have no knowledge about the upstream services that consume its resources.
- the social connectors should consume the sso settings service for retrieving the sso settings and then handle the login logic.